src/mito2/src/wal/entry_distributor.rs

```diff
@@ -0,0 +1,643 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::cmp::min;
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use api::v1::WalEntry;
+use async_stream::stream;
+use common_telemetry::{debug, error};
+use futures::future::join_all;
+use snafu::ensure;
+use store_api::logstore::entry::Entry;
+use store_api::logstore::provider::Provider;
+use store_api::storage::RegionId;
+use tokio::sync::mpsc::{self, Receiver, Sender, UnboundedReceiver, UnboundedSender};
+use tokio::sync::oneshot;
+use tokio_stream::StreamExt;
+
+use crate::error::{self, Result};
+use crate::wal::entry_reader::{decode_raw_entry, WalEntryReader};
+use crate::wal::raw_entry_reader::RawEntryReader;
+use crate::wal::{EntryId, WalEntryStream};
+
+/// [WalEntryDistributor] distributes Wal entries to specific [WalEntryReceiver]s based on [RegionId].
+pub(crate) struct WalEntryDistributor {
+    raw_wal_reader: Arc<dyn RawEntryReader>,
+    provider: Provider,
+    /// Sends [Entry] to receivers based on [RegionId]
+    senders: HashMap<RegionId, Sender<Entry>>,
+    /// Waits for the arg from the [WalEntryReader].
+    arg_receivers: Vec<(RegionId, oneshot::Receiver<EntryId>)>,
+}
+
+impl WalEntryDistributor {
+    /// Distributes entries to specific [WalEntryReceiver]s based on [RegionId].
+    pub async fn distribute(mut self) -> Result<()> {
+        let args_futures = self
```

Comment:
```suggestion
        let arg_futures = self
```
As `arg_receivers` is arg_xxx